stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+#include <stan/math/rev/functor/cvodes_utils.hpp>
+#include <stan/math/rev/functor/ode_store_sensitivities.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/functor/for_each.hpp>
+#include <cvodes/cvodes.h>
+#include <nvector/nvector_serial.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <algorithm>
+#include <ostream>
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Integrator interface for CVODES' ODE solvers (Adams & BDF
+ * methods).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_y0 Type of scalars for initial state
+ * @tparam T_param Type of scalars for parameters
+ * @tparam T_t0 Type of scalar of initial time point
+ * @tparam T_ts Type of time-points where ODE solution is returned
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args>
+class cvodes_integrator_adjoint_vari : public vari {
+  using T_Return = return_type_t<T_y0, T_t0, T_ts, T_Args...>;
+  using T_y0_t0 = return_type_t<T_y0, T_t0>;
+
+  const std::decay_t<F> f_;
+  arena_t<Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1>> y0_;
+  arena_t<T_t0> t0_;
+  std::vector<arena_t<T_ts>, arena_allocator<arena_t<T_ts>>> ts_;
+
+  double relative_tolerance_forward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_forward_;
+  double relative_tolerance_backward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_backward_;
+  double relative_tolerance_quadrature_;
+  double absolute_tolerance_quadrature_;
+  long int max_num_steps_;                  // NOLINT(runtime/int)
+  long int num_steps_between_checkpoints_;  // NOLINT(runtime/int)
+  int interpolation_polynomial_;
+  int solver_forward_;
+  int solver_backward_;
+
+  std::ostream* msgs_;
+
+  std::tuple<arena_t<
+      plain_type_t<decltype(deep_copy_vars(std::declval<const T_Args&>()))>>...>
+      local_args_tuple_;
+  std::tuple<arena_t<
+      plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>>...>
+      value_of_args_tuple_;
+
+  size_t N_;
+  bool forward_returned_;
+  bool backward_is_initialized_;
+
+  size_t num_t0_vars_;
+  size_t num_ts_vars_;
+  size_t num_y0_vars_;
+  size_t num_args_vars_;
+  size_t num_vars_;
+
+  arena_t<Eigen::VectorXd> state_forward_;
+  arena_t<Eigen::VectorXd> state_backward_;
+  arena_t<Eigen::VectorXd> quad_;
+
+  vari** non_chaining_varis_;
+
+  vari** t0_varis_;
+  vari** ts_varis_;
+  vari** y0_varis_;
+  vari** args_varis_;
+
+  int index_backward_;
+
+  struct cvodes_solver;
+  cvodes_solver* solver_;
+
+  static constexpr bool is_var_ts{is_var<T_ts>::value};
+  static constexpr bool is_var_t0{is_var<T_t0>::value};
+  static constexpr bool is_var_y0{is_var<T_y0_t0>::value};
+  static constexpr bool is_any_var_args{
+      disjunction<is_var<scalar_type_t<T_Args>>...>::value};
+  static constexpr bool is_var_return{is_var<T_Return>::value};
```

Comment:
underscores at the end for these 

 ```suggestion
  static constexpr bool is_var_return_{is_var<T_Return>::value};
```

Also can this ever be false?

Reply:
underscore... sure. Yes, this can be false if only doubles are passed in. The reason as to why the code still lives in rev is that we still need AD due to the stiff solver requiring the Jacobian.